src/edge_detection.cpp

```diff
@@ -1,3 +1,5 @@
+#pragma once
```

Comment:
I can explain the reason adding this :)
I wanted to keep the code the way it was - separated among `.h` and `.cpp` files, so it would provide some sort of convenience when reading it. I was following the 3rd method from the following [tutorial](https://www.codeproject.com/Articles/48575/How-to-define-a-template-class-in-a-h-file-and-imp) of how to keep both separate template's initialization and declaration. 
But obviously, it doesn't work, so I'll put everything into the header file.

Reply:
Putting everything in header file is not a good way as an user would not need to know the implementation. Let's me think a little about this...

Reply:
This is my idea:
1. Declare future helper class at the beginning of the header file:
```cpp
class EdgeDetectionHelper;
```
2. In header file create a base template class:
```cpp
template <typename _Type>
class EdgeDetectionBase
{
public:
    void find( const PenguinV_Image::Image & image, const EdgeParameter & edgeParameter = EdgeParameter() )
    {
        find( image, 0, 0, image.width(), image.height(), edgeParameter );
    }

    void find( const PenguinV_Image::Image & image, uint32_t x, uint32_t y, uint32_t width, uint32_t height, const EdgeParameter & edgeParameter = EdgeParameter() )
    {
        EdgeDetectionHelper::findEdgePoints( *this, image, x, y, width, height, edgeParameter )
    }

    const std::vector < Point2DBase<_Type> > & positiveEdge() const
    {
        return positiveEdgePoint;
    }
    const std::vector < Point2DBase<_Type> > & negativeEdge() const
    {
        return negativeEdgePoint;
    }

    friend class EdgeDetectionHelper;

private:
    std::vector < Point2DBase<_Type> > positiveEdgePoint;
    std::vector < Point2DBase<_Type> > negativeEdgePoint;
};
```
3. Create helper class with 2 static functions:
```cpp
class EdgeDetectionHelper
{
public:
    void findEdgePoints( EdgeDetectionBase<double> & edgeDetection, const PenguinV_Image::Image & image, uint32_t x, uint32_t y, uint32_t width, uint32_t height, const EdgeParameter & edgeParameter );
    void findEdgePoints( EdgeDetectionBase<float> & edgeDetection, const PenguinV_Image::Image & image, uint32_t x, uint32_t y, uint32_t width, uint32_t height, const EdgeParameter & edgeParameter );
};
```
4. Whole real implementation remains the in source file.